### ISSUE TITLE
The use of barrel files

### ISSUE THREAD

**Issue by s4m0r4m4** [nrwl/nx#3426]
## Description
I am using nx to manage multiple angular projects and angular component libraries. I have a library that includes angular components as well as useful typescript helper functions. I want to use those typescript helper functions in a webworker, but when I try to import the typescript helper function from the library barrel file, it breaks the worker compilation because the worker compiler doesn't know how to interpret the angular component's non-javascript content (HTML, etc.). 

I originally submitted something to the angular team (https://github.com/angular/angular-cli/issues/18206) and was basically told that barrel files are a bad idea for web workers - so instead I just do direct imports for web workers. That's fine and I can live with that if its just limited to the scope of webworkers. 

However, the angular team also said:

> barrel files are known to cause optimization bailouts.

References:
- https://github.com/angular/angular-cli/issues/18191#issuecomment-656031070
- https://github.com/angular/angular-cli/issues/16799

I guess I'm looking for some guidance here. Nx seems to promote barrel files for libs, but that seems to cause optimization issues? The [documentation](https://nx.dev/angular/tutorial/08-create-libs) is lacking when it comes to library design and what are the appropriate uses of barrel files.

**Comment by FrozenPandaz** [comment]
That's a good question that could use more documentation.

Barrel files in libraries define their public API. We want to hide certain implementation details of our lib and only expose the surface area that other consumer libs should be using. They are definitely recommended.

I think what causes issues with barrel files described in the Angular issue is when too many things that are actually unrelated are exported from a single barrel file.

Here's how this relates to Nx development.

1. How can we avoid barrel files that export unrelated things?

Because each lib has it's own barrel file, this is really a question of what code should be grouped in a lib and what should be split out into a separate lib.

2. What should be a separate lib?

One consideration that I use to decide whether to group some functionality in a lib rather than split it into separate libs is who should be depending on this lib.

For example, in early projects, it is common to create a `@my-proj/shared/ui` and drop all shared components into that one lib. However, then other projects need to import from `@my-proj/shared/ui` whenever they need any single component out of the shared components. This also causes slow build times and a lot of the same issues described in that issue.

In summary, I think the answer to your question is that barrel files are definitely recommended in Nx. But do keep those issues in mind when deciding to have one lib handle several unrelated code.

I'm going to relabel this as an enhancement for our docs so that we can add more information there.

**Comment by s4m0r4m4** [comment]
Thanks for clarifying, I appreciate it!

**Comment by nhhockeyplayer** [comment]
Im against barrels

Nx hasnt even set up a root path for the libs dir in PATHS in the tsconfig.base.json yet it expects us to operate automated code completion for imports across all languages?

my 45k line codebase and high end IDE webstorm refuses to articulate a single import other than packages.json

no documentation on whether we are suppose to setup our own file based packages to circumvent

the barrel file for a lib I created wont even work

Im dismayed and on 5th day of this wonde3rign when i will be able to build

this framework Nx is the savior for corporations and their massive libraries they operate

I want to fly and still on the ground

praying for a fix

in the interim Im still slugging it out with barrels, paths and all the mile long issues on this

I hope this doesn't come as disparaging
I want the best from the best

Nx is the best

lets make it the best

**Comment by nhhockeyplayer** [comment]
cannot import anything beneath my barrel file for my angular core library

this is appearing to be a land mine

5th day and still trying to build

if theres a way to toss barrel files please do so they lean towards javascripters

Nx is a few steps above yes?

this library eFrastructure is the bedrock of everything

and deserves immediate attention and high end solution

the issues are endless and still unresolved

**Comment by nhhockeyplayer** [comment]
I think barrel files need to be eliminated completely

Im not seeing any documentation that justifies Nx needs to be schematic'd with barrel files but only if i create a workspace lib I see an index file land and a function can someone explain what Nx needs from this and why its being schematic'd as such?

Ok so angular libs get schematic'd with an angular.module which qualifies as a legit es2015 module and it works imports fine self contained

we can create an angular module and export anything within that

and our angular modules behave like libraries

theres no backing documentation to argue the mandate for barrel files

the last thing I want to d after I model 45k lines of code is re-spell everything out in a barrel file thats already elegantly modularized in an angular module

as for isomorphic types shared full-stack client & server... if those types (type files which comply as an es2015 module fo classes abstract classes interfaces and custom types and enums) whats so bad about importing them at will FROM an angular module modeling the lib?

WHY does Nx need barrel files for the javascripters and react? and frameworks that dont model angular modules? If so then am I free to model all my libraries as an angular module?
Im finding myself being forced to operate barrel files with Nx and its very uncomfortable and bogging down development having to code an export for everything I already coded thats modularized and already exported.

Can someone shed some light here?

Can Nx live without barrel files? and if so why wont it?

Can I just be done with it and stick to angular modules and ditch all my barrel files?

Thanks for the light

**Comment by nhhockeyplayer** [comment]
at would be nice
is if the schematics asked what KIND of module we want to model
so we could have the option of ditching barrel files altogether

is it theoretically ok to model all out libraries (including types and just pure non-angular typescript) as an angular module?

are we free to break rank? or is Nx dependent internally somehow on these barrel files

would love to know if there is a hidden mystery with these barrel files and why they are being promoted when other means could automate and model something other than barrel files which are a bear to maintain

**Comment by yuchant** [comment]
So is the recommendation to use a lib per component? It's vastly helpful when a component has say a huge video player library attached, and it only get chunked into a page that requires it.

**Comment by nhhockeyplayer** [comment]
angular modules are ES2015 modules by spec

so a *single .module file is sufficient to gather up anything and everything PRIVATE/PUBLIC api beneath the lib folder and used as such and thats what the angular.module does

but thats not good enough the outside world need to impose its BAGGAGE onto clean angular devs

there is no need for a barrel file except to accomodate the non-angular development world
but the angular development world is mandated and forced to do barrel files which do not even hold up to paths and classname ambiguities when automatically generated to avoid massive maintenance nightmare. these mandates I speculate are coming out of the ng-packagr/compiler devs

how about since a MODULE is an es2015 module...lets ditch barrel files altogether and implement an automated MODULE scheme which defines the public/private(if any) api

barrel files have been a terrible scourge everywhere and need to be eradicated and automated

**Comment by nhhockeyplayer** [comment]
barrel files are caveman javascript and bad design no one ever took into account managing them nor their size nor scale nor impact
and this design should be removed from everything

https://adrianfaciu.dev/posts/barrel-files/

>There are people arguing that one does not need to use barrel files, at least inside Angular applications since we have [NgModule](https://angular.io/api/core/NgModule). Even the [official documentation](https://angular.io/guide/glossary#barrel) hints towards this.

and its been a scourge pinning down elite developers in build issues

an angular module is an es2015 module as is everything the field does not need barrel files to re-articulate interfaces that are clearly standardized by the angular module. And now Victors masterpiece is affected by the angular compilers inability to compile paths producing false ambiguities
angular compiler team is now all in on barrel files now
this circumvents any remedy

wish I had better to say
